Perf attrib plots

vikram-narayan opened the pull request:
Plot perf attribution:
- cumulative returns
- alpha returns
- individual factors

Screenshots of plots
![returns_attribution](https://user-images.githubusercontent.com/8713773/29798636-2e25a48e-8c2c-11e7-90c1-992c856df01c.png)
![alphas](https://user-images.githubusercontent.com/8713773/29798637-2e34efe8-8c2c-11e7-9759-87e39003b789.png)
![returns](https://user-images.githubusercontent.com/8713773/29798638-2e37e004-8c2c-11e7-96bb-af953d1e0783.png)

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -16,6 +16,7 @@
 import pandas as pd
 
 import empyrical
```
I believe @twiecki prefers `import empyrical as ep` to keep API compatibility

twiecki commented:
This looks great. Can you show what the plots look like?

twiecki commented:
@richafrank Ready for engineering review.

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    ax.plot(specific_returns, color='b', label='Cumulative specific returns')
+    ax.plot(common_returns, color='r', label='Cumulative common returns')
+
+    ax.set_title('Time Series of Cumulative Returns')
```
We prefer to have all titles, labels and legends have only the first letter of the first word capitalized, to keep plotting style standardized across tear sheets. Refer to #422

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    return ax
+
+
+def plot_alpha_returns(alpha_returns, ax=None):
```
It's important to `return ax` in any function that plots stuff.

Each of pyfolio's tear sheets should be one single image: that is, if I saved it the whole tear sheet should save as a single png, not one png for each plot in the tear sheet. This is to make it easier for the analysis to be saved, passed around, etc. We go to great lengths in `tears.py` to achieve this, and the way we do it is by making a single `matplotlib` figure with several axes, and pass these axes to functions in other modules: `tears.py` then expects the axes back, with the stuff having been plotted.

We also don't lose the "just plot" functionality with this method either, because if a user just calls `plot_foo` and _doesn't_ pass in an axes, then `ax=None`, `plt.gca()` and the `ax` that gets returned is just thrown away, leaving only the plotted output to be shown with `plt.show()`. So we don't really lose anything here.

There are some other small necessary changes that I'll outline in separate comments. Hope this made sense!

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    if ax is None:
+        ax = plt.gca()
+
+    plt.hist(alpha_returns, color='g', label='multi-factor alpha')
```
Related to the `return ax` comment: this needs to be `ax.hist`

Also, for style, `label='Multi-factor alpha'` (capitalization)

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+        ax = plt.gca()
+
+    plt.hist(alpha_returns, color='g', label='multi-factor alpha')
+    plt.title('Histogram of alphas (specific pnl / AUM)')
```
`ax.set_title`

Also, did we agree to use AUM or GMV? @twiecki

twiecki replied:
This is based on returns, so we can just drop what's in the parantheses.

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+
+    plt.hist(alpha_returns, color='g', label='multi-factor alpha')
+    plt.title('Histogram of alphas (specific pnl / AUM)')
+    plt.axvline(0, color='k', linestyle='--', label='zero')
```
`ax.axvline`
`label='Zero'` (capitalization for style)

twiecki replied:
Yes, these calls all need to be converted to `ax.method`

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    plt.axvline(0, color='k', linestyle='--', label='zero')
+
+    avg = alpha_returns.mean()
+    plt.axvline(avg, color='b', label='mean = {: 0.5f}'.format(avg))
```
`ax.axvline`
`label='Mean = ...` (capitalization)

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    plt.legend()
+
+
+def plot_factor_contribution_to_perf(perf_attrib_data, ax=None):
```
`return ax` here too

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    ax.stackplot(perf_attrib_data.index, perf_attrib_data.iloc[:, :-3])
+
+    ax.axhline(0, color='k')
+    ax.legend(frameon=True, loc=2)
```
`loc='upper left'` for readability

Also, `frameon=True, framealpha=0.5` for style (see #422)

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    ax.axhline(0, color='k')
+    ax.legend(frameon=True, loc=2)
+
+    ax.set_ylabel('Contribution to Returns by Factor')
```
`'Contribution to returns by factor'` (capitalization)

eigenfoo commented on pyfolio/perf_attrib.py:
```diff
@@ -118,21 +130,83 @@ def create_perf_attrib_stats(perf_attrib):
+    ax.legend(frameon=True, loc=2)
+
+    ax.set_ylabel('Contribution to Returns by Factor')
+    ax.set_title('Returns Attribution', fontsize='large')
```
`'Returns attribution'` (capitalization)

Also, don't use `fontsize='large'`. Avoid manually setting font sizes for style.

twiecki commented:
Thanks for the comprehensive review @eigenfoo!

vikram-narayan commented:
@eigenfoo thanks for the review, updated the PR based on your feedback
@twiecki sorry didn't see your request for plot images, updated the PR description

eigenfoo commented:
I might be missing some context, but would it not be more useful to have a cumulative returns attribution, instead of a returns attribution (for the first plot)? That is, simply do a `df.cumsum` before plotting.

twiecki commented on pyfolio/perf_attrib.py:
```diff
@@ -112,27 +124,97 @@ def create_perf_attrib_stats(perf_attrib):
+    if ax is None:
+        ax = plt.gca()
+
+    ax.plot(returns, color='g', label='Total returns')
```
You need to call `ep.cum_returns` on all these to convert them to cumulative returns. Also change the label of `Total returns` to `Cumulative total returns`.

vikram-narayan replied:
fixed

twiecki commented on pyfolio/perf_attrib.py:
```diff
@@ -112,27 +124,97 @@ def create_perf_attrib_stats(perf_attrib):
+
+    Parameters
+    ----------
+    perf_attrib_data : pd.DataFrame
```
missing doc for exposures.

vikram-narayan replied:
fixed

twiecki commented on pyfolio/perf_attrib.py:
```diff
@@ -112,27 +124,97 @@ def create_perf_attrib_stats(perf_attrib):
+    Parameters
+    ----------
+    perf_attrib_data : pd.DataFrame
+        df with factors, common returns, and specific returns as columns,
```
Need an example dataframe of what this should look like.

vikram-narayan replied:
fixed

twiecki commented:
I also tried to call a plotting function but got:
```python
pf.perf_attrib.plot_factor_contribution_to_perf(perf_attrib_pf)

ValueErrorTraceback (most recent call last)
<ipython-input-17-75b93317bdd3> in <module>()
----> 1 pf.perf_attrib.plot_factor_contribution_to_perf(perf_attrib_pf)

/home/wiecki/.local/lib/python2.7/site-packages/pyfolio/perf_attrib.pyc in plot_factor_contribution_to_perf(perf_attrib_data, ax)
    204         ax = plt.gca()
    205 
--> 206     ax.stackplot(perf_attrib_data.index, perf_attrib_data.iloc[:, :-3])
    207 
    208     ax.axhline(0, color='k')

/opt/miniconda/lib/python2.7/site-packages/matplotlib/__init__.pyc in inner(ax, *args, **kwargs)
   1890                     warnings.warn(msg % (label_namer, func.__name__),
   1891                                   RuntimeWarning, stacklevel=2)
-> 1892             return func(ax, *args, **kwargs)
   1893         pre_doc = inner.__doc__
   1894         if pre_doc is None:

/opt/miniconda/lib/python2.7/site-packages/matplotlib/axes/_axes.pyc in stackplot(self, x, *args, **kwargs)
   4596     @unpack_labeled_data(replace_all_args=True, label_namer=None)
   4597     def stackplot(self, x, *args, **kwargs):
-> 4598         return mstack.stackplot(self, x, *args, **kwargs)
   4599     stackplot.__doc__ = mstack.stackplot.__doc__
   4600 

/opt/miniconda/lib/python2.7/site-packages/matplotlib/stackplot.pyc in stackplot(axes, x, *args, **kwargs)
    111     coll = axes.fill_between(x, first_line, stack[0, :],
    112                              facecolor=color, label=six.next(labels, None),
--> 113                              **kwargs)
    114     coll.sticky_edges.y[:] = [0]
    115     r = [coll]

/opt/miniconda/lib/python2.7/site-packages/matplotlib/__init__.pyc in inner(ax, *args, **kwargs)
   1890                     warnings.warn(msg % (label_namer, func.__name__),
   1891                                   RuntimeWarning, stacklevel=2)
-> 1892             return func(ax, *args, **kwargs)
   1893         pre_doc = inner.__doc__
   1894         if pre_doc is None:

/opt/miniconda/lib/python2.7/site-packages/matplotlib/axes/_axes.pyc in fill_between(self, x, y1, y2, where, interpolate, step, **kwargs)
   4785 
   4786         if not (x.shape == y1.shape == y2.shape == where.shape):
-> 4787             raise ValueError("Argument dimensions are incompatible")
   4788 
   4789         mask = reduce(ma.mask_or, [ma.getmask(a) for a in (x, y1, y2)])

ValueError: Argument dimensions are incompatible
```

twiecki commented:
We also need an area-plot of risk exposures over time.

vikram-narayan commented:
@twiecki made a change to `plot_factor_contribution_to_perf` which should fix your issue.

As we discussed, I'm planning to add the area-plot of risk exposures in a subsequent PR (let me know if you want that in this PR instead)

twiecki commented:
I think this looks good @vikram-narayan. Still requires some touch-ups but we can do that in the next PR.

twiecki commented:
Oops, I guess you should have done the merge, sorry about that.

vikram-narayan commented:
np, working on the other PR now
